design-proposals/persist-vmi-firmware-uuid.md

```diff
@@ -0,0 +1,152 @@
+By storing the firmware UUID, we ensure that it remains consistent across VMI restarts.
+which is crucial for applications and services that rely on the UUID for identification or licensing purposes.
+
+## Motivation
+By definition, UUID must be a universally-unique identifier.
+The current implementation for the automatically-computed Firmware UUID is not universally unique: 
+Two VMs with the same name get the same UUID, even if they reside in different namespaces or different clusters.
+This proposal discusses how to fix this, while ensuring that existing VMs keep their current ID.
+
+## Goals
+Generate and persist universally-unique Firmware.UUID for new VMs
+Keep the Firmware.UUID of pre-existing VMs
+
+## Non Goals
+
+
+## Definition of Users
+End Users: Individuals or organizations running VMs and VMIs on KubeVirt who require consistent firmware UUIDs for their applications.
+
+## User Stories
+As an end-user, I expect my VMI to maintain its identity across restarts.
+
+## Repos
+Kubevirt/kubevirt
+
+# Design
+
+The introduction of a persistent, universally unique firmware UUID will proceed in two phases:
+
+### Phase 1: Preserve Current Generation Method and Persist UUID
+
+In this phase, we aim to minimize workload disruption by persisting the UUID generated with the current method. The goal is to ensure that VMs retain their UUID across restarts without changing the existing generation approach. Below are the potential solutions for this phase:
+
+1. **Persist UUID in the VM’s Spec Field**  
+   **Description:** The UUID would be generated upon the VM’s first boot and stored in the VM’s spec field.
+
+   **Pros:**
+   - Straightforward to implement.
+   - Avoids adding a new API field, fully backward compatible.
+
+   **Cons:**
+   - This approach "abuses" the spec by storing data that does not represent the user's desired state.
+   - Could lead to cluttered and harder-to-read VM definitions.
+
+
+2. **Generate UUID via Webhook and Store in the Spec Field**  
+   **Description:** If no UUID is specified by the user, a webhook generates a random UUID and assigns it to the VM’s spec field.
+
+   **Pros:**
+   - Similar to the first option (no new API field, backward compatible).
+
+   **Cons:**
+   - Adds complexity by involving webhooks, which could slow down performance if numerous VMs start simultaneously.
+   - May potentially bottleneck `virt-api`.
+
+
+3. **Add a Firmware UUID Field to VM Status**  
+   **Description:** Store the generated UUID within the VM’s status.
+
+   **Pros:**
+   - Keeps the spec clean.
+   - Focuses on VM status for generated information.
+
+   **Cons:**
+   - Could introduce a new API field, affecting long-term maintenance and adding to potential API bloat.
+   - The status field is dynamic and frequently updated by controllers, which could make the stored UUID fragile or prone to accidental modification.
+
```

Comment:
5.
* Just before kubevirt is upgraded, persist Firmware.UUID of existing VMs in vm.spec. From that point on, a VM without Firmware.UUID is new.
* From that point on, use vm.metadata.uid if vm.spec.template.spec.Firmware.UUID is not defined.

Pros:
* The upgrade-specific code can be deleted after two or three releases
* Very simple logic post-upgrade
* Disturbance to git-ops is limited to pre-existing (somewhat) buggy VMs.

Cons:
* We need to add a piece of code (to virt-operator?) to implement the upgrade-specific piece.

Reply:
> then we should probably wait for 1-2 releases until all users upgraded their machine, and we will be free to change the generation logic as we pleases

Not sure I follow your comment. I'm suggesting to do a one-off persisting of buggy uuid during KubeVirt upgrade. The new logic (`vm.metadata.uid`) takes effect immediately after that.

Reply:
understood, my bad

Reply:
@dankenigsberg Patching the spec of a running VM will raise the required restart condition.